src/zcash/Proof.cpp

```diff
@@ -0,0 +1,252 @@
+#include "Proof.hpp"
+
+#include <boost/static_assert.hpp>
+
+#include "crypto/common.h"
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+
+using namespace libsnark;
+
+typedef alt_bn128_pp curve_pp;
+typedef alt_bn128_pp::G1_type curve_G1;
+typedef alt_bn128_pp::G2_type curve_G2;
+typedef alt_bn128_pp::GT_type curve_GT;
+typedef alt_bn128_pp::Fp_type curve_Fr;
+typedef alt_bn128_pp::Fq_type curve_Fq;
+typedef alt_bn128_pp::Fqe_type curve_Fq2;
+
+BOOST_STATIC_ASSERT(sizeof(mp_limb_t) == 8);
+
+namespace libzcash {
+
+// Compares two bigints, returning 0 if equal, 1 if a > b, and -1 if a < b
+int cmp_bigint(const bigint<4> &a, const bigint<4> &b)
+{
+    for (ssize_t i = 3; i >= 0; i--) {
+        if (a.data[i] < b.data[i]) {
+            return -1;
+        } else if (a.data[i] > b.data[i]) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+// Returns whether a > b in lexicographic order
+bool cmp_fq2(const curve_Fq2 &a, const curve_Fq2 &b)
+{
+    auto cmp = cmp_bigint(a.c1.as_bigint(), b.c1.as_bigint());
+    if (cmp == 0) {
+        return cmp_bigint(a.c0.as_bigint(), b.c0.as_bigint()) > 0;
+    }
+
+    return cmp > 0;
+}
+
+// Writes a bigint in big endian
+template<mp_size_t LIMBS>
+void write_bigint(unsigned char *ptr, const bigint<LIMBS> &val)
```

Comment:
It would be nice if there was a way to locally guarantee memory safety here. Everywhere we call this we're calling it on a base_blob's `.data()` so maybe pass the `base_blob` in and check against its size?